File: python/tutorials/algo/sort/quick_sort.py
"""
quick sort, worst case: O^2, but it's rare. mostly it's nlog(n)

https://en.wikipedia.org/wiki/Quicksort

"""
import sys
import random
def patition(arr, lo, hi):
    pivot = arr[hi]
    i = lo
    for j in range(lo, hi):
        if arr[j] < pivot:
            arr[i],arr[j] = arr[j], arr[i]
            i = i+1
    arr[i], arr[hi] = arr[hi],arr[i]
    return i

#Lomuto partition scheme to select a pivot
def quick_sorting(arr, lo, hi):
    if lo < hi:
        p = patition(arr,lo,hi)
        quick_sorting(arr,lo, p-1)
        quick_sorting(arr, p+1, hi)



#02/06/19, use the first element as pivot
def partition2(arr, low, high):
    pivot = arr[low]
    i = high+1
    for j in range(high,low,-1):
        if arr[j] > pivot:
            i = i - 1
            arr[j],arr[i] = arr[i], arr[j]

    arr[low], arr[i-1] = arr[i-1],arr[low]

    return i-1

def quick_sorting2(arr,lo,hi):
    if lo < hi:
        p = partition2(arr,lo, hi)
        quick_sorting2(arr, lo, p-1)
        quick_sorting2(arr, p+1, hi)



def main(argv):
    #arr = random.sample(range(0,100),5)
    arr = [77, 46, 90, 79, 73]
    print(arr)
    quick_sorting2(arr, 0, len(arr)-1)
    print(arr)

if __name__ == '__main__':
    main(sys.argv[1:])

File: java/java_tests/src/test/java/com/entrepidea/algo/data_structure/sort/QuickSort.java
package com.entrepidea.algo.data_structure.sort;

import org.junit.Test;

import java.util.concurrent.ThreadLocalRandom;

/**
Quick sorting: the idea of quick sorting is to pick up a random element (known as "pivot"), 
to compare with the rest. Those less than the pivot form a new array, those greater forms another, 
these two as well as the pivot will form an array like less_array + pivot + greater_array. 
The above steps are repeated against those less and greater respectively (recursion).

**/
public class QuickSort {

	//non-recursive implementation
	private int[] concatenate(int[] less, int pivot, int[] greater) {
		int[] ret = new int[less.length + greater.length + 1];
		for (int i = 0; i < less.length; i++) {
			ret[i] = less[i];
		}
		ret[less.length] = pivot;
		for (int i = 0; i < greater.length; i++) {
			ret[less.length + 1 + i] = greater[i];
		}
		return ret;
	}

	private int[] sort(int[] array) {
		int[] ret = array;
		if (array.length <= 1) {
			return ret;
		} else {
			// select and remove a pivot
			int pivot_index = array.length / 2;
			int pivot = array[pivot_index];
			int[] less = new int[array.length];
			int[] greater = new int[array.length];
			int index = 0, index1 = 0, index2 = 0;
			for (int x : array) {
				if (index == pivot_index) {
					index++;
					continue; // skip the pivot element
				}
				if (x <= pivot) {
					less[index1++] = x;
				} else {
					greater[index2++] = x;
				}
				index++;
			} // ~for~
			int[] new_less = new int[index1];
			for (int i = 0; i < index1; i++) {
				new_less[i] = less[i];
			}
			int[] new_greater = new int[index2];
			for (int i = 0; i < index2; i++) {
				new_greater[i] = greater[i];
			}
			less = null;
			greater = null;
			ret = concatenate(sort(new_less), pivot, sort(new_greater));
			return ret;
		}
	}

	@Test
	public void testSorting() {

		int[] testArray = ThreadLocalRandom.current().ints(0,100).limit(10).toArray();


		for (int i : testArray) {
			System.out.print(i + ",");
		}
		System.out.println();

		int[] ret = new QuickSort().sort(testArray);
		for (int i : ret) {
			System.out.print(i + ",");
		}
		System.out.println();
	}


	//recursive implementation
	//Lomuto partition scheme to select a pivot
	//below code follows the psudo code from https://en.wikipedia.org/wiki/Quicksort
	private int partition(int[] arr, int lo, int hi){
		int pivot = arr[hi];
		int i = lo;

		for( int j=lo; j<hi;j++){
			if (arr[j]<pivot){
				int temp = arr[i];
				arr[i] = arr[j];
				arr[j] = temp;
				i++;
			}
		}
		int temp = arr[i];
		arr[i] = arr[hi];
		arr[hi] = temp;
		return i;
	}

	private void sort(int[] arr, int lo, int hi){
		if(lo<hi){
			int p = partition(arr, lo, hi);
			sort(arr, lo, p-1);
			sort(arr, p+1, hi);
		}
	}

	@Test
	public void testSorting2(){
		int[] arr = ThreadLocalRandom.current().ints(0,100).limit(20).toArray();
		sort(arr, 0, arr.length-1);
		for (int a : arr){
			System.out.print(a);
			System.out.print("\t");
		}
		System.out.println();
	}

}

